File: old_code/src/utils/ParserPath.ts
/**
 * Represents a parser path during parsing.
 */
export class ParserPath {
  private parserPathId: number;
  private lexerPathId: number;
  private activeProductions: Production[];
  private activeMatches: ProductionMatch[];
  private position: number;

  /**
   * Creates a new ParserPath instance.
   * @param parserPathId The ID of the parser path
   * @param lexerPathId The ID of the associated lexer path
   * @param position The current position in the input
   */
  constructor(parserPathId: number, lexerPathId: number, position: number) {
    this.parserPathId = parserPathId;
    this.lexerPathId = lexerPathId;
    this.activeProductions = [];
    this.activeMatches = [];
    this.position = position;
  }

  /**
   * Gets the parser path ID.
   */
  public getParserPathId(): number {
    return this.parserPathId;
  }

  /**
   * Gets the lexer path ID.
   */
  public getLexerPathId(): number {
    return this.lexerPathId;
  }

  /**
   * Gets the active productions.
   */
  public getActiveProductions(): Production[] {
    return this.activeProductions;
  }

  /**
   * Adds an active production.
   * @param production The production to add
   */
  public addActiveProduction(production: Production): void {
    this.activeProductions.push(production);
  }

  /**
   * Gets the active matches.
   */
  public getActiveMatches(): ProductionMatch[] {
    return this.activeMatches;
  }

  /**
   * Adds an active match.
   * @param match The match to add
   */
  public addActiveMatch(match: ProductionMatch): void {
    this.activeMatches.push(match);
  }

  /**
   * Gets the current position.
   */
  public getPosition(): number {
    return this.position;
  }

  /**
   * Sets the current position.
   * @param position The new position
   */
  public setPosition(position: number): void {
    this.position = position;
  }

  /**
   * Sets the parser path ID.
   * @param id The new parser path ID
   */
  public setParserPathId(id: number): void {
    this.parserPathId = id;
  }

  /**
   * Sets the lexer path ID.
   * @param id The new lexer path ID
   */
  public setLexerPathId(id: number): void {
    this.lexerPathId = id;
  }

  /**
   * Creates a string representation of this parser path.
   */
  public toString(): string {
    return `ParserPath(id: ${this.parserPathId}, lexerPath: ${this.lexerPathId}, position: ${this.position})`;
  }
}

// Import required classes
import { Production } from './Production';
import { ProductionMatch } from './ProductionMatch';

File: src/StepParser/Core/ParserPath.cs
using StepParser.Core;
using StepParser.Memory;

namespace StepParser.Core;

/// <summary>
/// Represents a parser path for tracking parsing state.
/// </summary>
public class ParserPath : IPoolableObject
{
    private int _parserPathId;
    private int _lexerPathId;
    private int _position;
    private readonly List<Production> _activeProductions;
    private readonly List<ProductionMatch> _activeMatches;
    private bool _isInUse;

    /// <summary>
    /// Initializes a new instance of the <see cref="ParserPath"/> class.
    /// </summary>
    /// <param name="parserPathId">The parser path ID.</param>
    /// <param name="lexerPathId">The lexer path ID.</param>
    /// <param name="position">The position in the input.</param>
    public ParserPath(int parserPathId, int lexerPathId, int position)
    {
        _parserPathId = parserPathId;
        _lexerPathId = lexerPathId;
        _position = position;
        _activeProductions = new List<Production>();
        _activeMatches = new List<ProductionMatch>();
        _isInUse = false;
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ParserPath"/> class with default values.
    /// </summary>
    public ParserPath() : this(0, 0, 0)
    {
    }

    /// <summary>
    /// Gets or sets the parser path ID.
    /// </summary>
    public int ParserPathId
    {
        get => _parserPathId;
        set => _parserPathId = value;
    }

    /// <summary>
    /// Gets or sets the lexer path ID.
    /// </summary>
    public int LexerPathId
    {
        get => _lexerPathId;
        set => _lexerPathId = value;
    }

    /// <summary>
    /// Gets or sets the position in the input.
    /// </summary>
    public int Position
    {
        get => _position;
        set => _position = value;
    }

    /// <summary>
    /// Gets the active productions.
    /// </summary>
    public List<Production> ActiveProductions => _activeProductions;

    /// <summary>
    /// Gets the active matches.
    /// </summary>
    public List<ProductionMatch> ActiveMatches => _activeMatches;

    /// <inheritdoc/>
    public bool IsInUse
    {
        get => _isInUse;
        set => _isInUse = value;
    }

    /// <summary>
    /// Adds an active production.
    /// </summary>
    /// <param name="production">The production to add.</param>
    public void AddActiveProduction(Production production)
    {
        _activeProductions.Add(production);
    }

    /// <summary>
    /// Adds an active match.
    /// </summary>
    /// <param name="match">The match to add.</param>
    public void AddActiveMatch(ProductionMatch match)
    {
        _activeMatches.Add(match);
    }

    /// <inheritdoc/>
    public void Reset()
    {
        _parserPathId = 0;
        _lexerPathId = 0;
        _position = 0;
        _activeProductions.Clear();
        _activeMatches.Clear();
        _isInUse = false;
    }

    /// <summary>
    /// Returns a string representation of this parser path.
    /// </summary>
    /// <returns>A string representation.</returns>
    public override string ToString()
    {
        return $"ParserPath(Id={_parserPathId}, LexerId={_lexerPathId}, Pos={_position}, Productions={_activeProductions.Count}, Matches={_activeMatches.Count})";
    }
}